bidict/_common.py

```
@@ -53,103 +102,164 @@ def __inverted__(self):
-    values = lambda self: self._inv.keys()
+        if valexists:
+            if val_clbhv is RAISE:
+                # since multiple values can have the same hash value, refer
+                # to the existing value via `_fwd[oldkey]` rather than `val`
+                raise ValueExistsError((oldkey, _fwd[oldkey]))
+            elif val_clbhv is IGNORE:
+                return
+        _fwd.pop(oldkey, None)
+        _inv.pop(oldval, None)
+        _fwd[key] = val
+        _inv[val] = key
+
+    def _update(self, key_clbhv, val_clbhv, *args, **kw):
+        if not args and not kw:
+            return
+
+        _fwd, _inv, _missing = self._fwd, self._inv, self._missing
+
+        # take fast path if passed only another bidict
+        if not kw and args and isinstance(args[0], BidirectionalMapping):
+            l = len(args)
+            if l != 1:
+                raise TypeError('Expected at most 1 argument, got %d' % l)
+            updatefwd = self._dcls(args[0]._fwd)
+            updateinv = self._dcls(args[0]._inv)
+        else:
+            if key_clbhv is RAISE:
+                items = frozenset(pairs(*args, **kw))
+                updatefwd = self._dcls(items)
+                if len(items) > len(updatefwd):
+                    raise NonuniqueKeysError(items)
+            else:
+                updatefwd = self._dcls(*args, **kw)
+            updateinv = self._dcls(inverted(updatefwd))
+            if len(updatefwd) > len(updateinv):
+                if val_clbhv is RAISE:
+                    raise NonuniqueValuesError(updatefwd)
+                updatefwd = self._dcls(inverted(updateinv))
+
+        common_vals = viewkeys(updateinv) & viewkeys(_inv)
+        if common_vals and val_clbhv is RAISE:
+            raise ValuesExistError(common_vals)
+
+        common_keys = viewkeys(updatefwd) & viewkeys(_fwd)
+        if common_keys and key_clbhv is RAISE:
+            raise KeysExistError(common_keys)
```

@lordmauve I realized with the version you reviewed, if `_update` was called with a RAISE collision behavior and given an existing item, it would raise rather than ignore it. I think it should raise if only the key and not the value, or only the value and not the key, of a new item collides with that of an existing item. So I took another pass at implementing this and I think now it has the correct and desired behavior. Could possibly be optimized (as well as more DRY), but thanks to some optimizations I realized I could do to `bidict.util.pairs`, it's still faster than the version on master, and safer to boot. Look forward to your review of this latest version, whenever you get a chance.